src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -162,6 +166,32 @@ static SemanticTokensHelpers()
             return (ComputeTokens(text.Lines, updatedClassifiedSpans, tokenTypesToIndex), isFinalized);
         }
 
+        private static async Task GetClassifiedSpansForDocumentAsync(
+            Document document,
+            TextSpan textSpan,
+            ArrayBuilder<ClassifiedSpan> classifiedSpans,
+            bool isRazorDoc,
+            CancellationToken cancellationToken)
+        {
+            var classificationService = document.GetRequiredLanguageService<IClassificationService>();
+
+            // Case 1 - C# and VB documents:
+            //     In C#/VB, the syntax classifier runs on the client. This means we only need to return semantic
+            //     classifications.
+            // Case 2 - Generated Razor documents:
+            //     In Razor, the C# syntax classifier does not run on the client. This means we need to return both
+            //     syntactic and semantic classifications.
+            //
+            // Ideally, Razor will eventually run the classifier on their end so we can get rid of this special
+            // casing: https://github.com/dotnet/razor-tooling/issues/5850
+            // Until then, the `AddSemanticClassificationsAsync` call below is special-cased to deal with both scenarios.
+            await SemanticClassificationCacheUtilities.AddSemanticClassificationsAsync(
+                document, textSpan, classificationService, classifiedSpans, isRazorDoc, cancellationToken).ConfigureAwait(false);
```

Comment:
so instead of passing isRazorDoc here (and then having AddSemandticClassificationsAsync do odd stuff (like adding syntactic classifications), i would instead have the code *here* fork it's behavior (with the good comment you have), and say htat in the razor case we get all diagnostics, where for all otehrs we just get semantic classification.